### PR TITLE
Update Perl doc for new database framework

### PR DESCRIPTION
**Pull request by nicolasbrossard**
This PR corrects the documentation errors found when merging minor into major and also adds the markdown files for the new database framework.

### CODE REVIEW

File: dicom-archive/updateMRI_Upload.pl [hunk 1]
```diff
@@ -12,7 +12,7 @@ =head1 NAME
```
Comment (cmadjar) [inline dicom-archive/updateMRI_Upload.pl]: missing updated MD file

File: uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm [hunk 2]
```diff
@@ -13,7 +13,7 @@ utilities
   use NeuroDB::ProcessingUtility;
```
Comment (cmadjar) [inline uploadNeuroDB/NeuroDB/MRIProcessingUtility.pm]: Missing updated MD file

File: uploadNeuroDB/NeuroDB/MRI.pm [hunk 3]
```diff
@@ -449,13 +449,13 @@ sub getObjective
```
Comment (cmadjar) [inline uploadNeuroDB/NeuroDB/MRI.pm]: missing updated MD file

File: docs/scripts_md/Database.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        die sprintf(
+            "SELECT on candidate table failed: %s (error code=%d)",
+            $e->errorMessage,
+            $e->errorCode
+        );
+    }
+
+# DESCRIPTION
+
+This class provides the basic SELECT, INSERT, UPDATE and DELETE methods
```
Comment (cmadjar) [inline docs/scripts_md/Database.md]: Nitpicking, should `SELECT`, `INSERT`, `UPDATE` and `DELETE` be in code format?

File: docs/scripts_md/Database.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+for all object brokers. The methods of this class should only be used by
+the object brokers themselves (except 'new' for creating a new database
+instance and 'connect'). Scripts and 'non-broker' classes that need access
```
Comment (cmadjar) [inline docs/scripts_md/Database.md]: same for `new`, `connect` maybe

File: docs/scripts_md/Database.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+to the database should rely on an appropriate object broker class to handle
+the requests.
+
+## Methods
+
+### `new(userName => $u, databaseName => $d, hostname => $h, password => $pwd, port =>$port)`  (constructor)
+
+Create a new instance of this class, without actually trying to connect
+to the database specified. All parameters are required except 'port', which
```
Comment (cmadjar) [inline docs/scripts_md/Database.md]: same for `port`

File: docs/scripts_md/Database.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+defaults to 3306 if not specified. If the user name, database name or host
+name are the empty string, the constructor will call `die`.
+
+INPUTS: a set of properties for the current database:
+
+\- name of the user for the (upcoming) connection
```
Comment (cmadjar) [inline docs/scripts_md/Database.md]: bad formatting in the MD file of the list. Ends up all in one row:

\- name of the user for the (upcoming) connection - name of the database - name of the host on which the database resides - password for the (upcoming) connection - port used for the (upcoming) connection (defaults to 3306 if not provided)

File: docs/scripts_md/Database.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+\- name of the host on which the database resides
+\- password for the (upcoming) connection
+\- port used for the (upcoming) connection (defaults to 3306 if not provided)
+
+RETURN: new instance of this class.
+
+### `connect()`
+
+Attempts to connect to the database using the connection parameters passed
+at construction time. This method will throw a DatabaseException if the
```
Comment (cmadjar) [inline docs/scripts_md/Database.md]: DatabaseException should be in code format

File: docs/scripts_md/Database.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+connection could not be established.
+
+### `pselect($query, @args)`
+
+Executes a select query on the database. This method will first `prepare`
```
Comment (cmadjar) [inline docs/scripts_md/Database.md]: select should be in code format and all cap

File: docs/scripts_md/Database.md [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+the statement passed as parameter before sending the request to the database.
+
+INPUTS: 
+    - select query to execute (containing the argument placeholders if any)
```
Comment (cmadjar) [inline docs/scripts_md/Database.md]: same here for select

File: docs/scripts_md/Database.md [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    - list of arguments to replace the placeholders with.
+
+RETURN: a reference to the array of records found. Each record is in fact a
+        reference to the list of values for the columns selected
```
Comment (cmadjar) [inline docs/scripts_md/Database.md]: missing dot at the end of the sentence

File: docs/scripts_md/Database.md [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+### `insertOne($tableName, $valuesRef)`
+
+Inserts one record in a given database table with the specified column values.
+This method will throw a DatabaseException if the record cannot be inserted.
```
Comment (cmadjar) [inline docs/scripts_md/Database.md]: DatabaseException in code format

File: docs/scripts_md/Database.md [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+### `insert($tableName, $columnNamesRef, $valuesRef)`
+
+Inserts one record in a given database table with the specified column values.
+This method will throw a `DatabaseException` if the record cannot be inserted.
+
+INPUTS: 
+    - name of the table in which to insert the record
+    - reference to an array containing the names of the columns whose values
+      will be modified by this `insert` statement.
```
Comment (cmadjar) [inline docs/scripts_md/Database.md]: insert all cap?

File: docs/scripts_md/MriScanTypeOB.md [hunk 14]
```diff
@@ -0,0 +1,95 @@
+# NAME
+
+NeuroDB::objectBroker::MriScanTypeOB -- An object broker for mri\_scan\_type records
```
Comment (cmadjar) [inline docs/scripts_md/MriScanTypeOB.md]: mri_scan_type should be in code format

Comment (cmadjar) [inline docs/scripts_md/MriScanTypeOB.md]: mri_scan_type should be in code format

File: docs/scripts_md/MriScanTypeOB.md [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+### new(db => $db) >> (constructor)
+
+Create a new instance of this class. The only parameter to provide is the
+`Database` object used to access the database.
+
+INPUT: the database object used to query the `mri_scan_type` table.
+
+RETURN: new instance of this class.
+
+### get($isCount, $columnValuesRef))
```
Comment (cmadjar) [inline docs/scripts_md/MriScanTypeOB.md]: missing a line space between title of the function and its description

File: docs/scripts_md/MriScanTypeOB.md [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+Fetches the entries in the `mri_upload` table that have specific column
+values. This method throws a `NeuroDB::objectBroker::ObjectBrokerException`
+if the operation could not be completed successfully.
```
Comment (cmadjar) [inline docs/scripts_md/MriScanTypeOB.md]: missing line space between end of descriptions and INPUTS

File: docs/scripts_md/MriScanTypeOB.md [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+
+Create a new instance of this class. The only parameter to provide is the
+`Database` object used to access the database.
+
+INPUT: the database object used to query the `mri_scan_type` table.
+
+RETURN: new instance of this class.
+
+### get($isCount, $columnValuesRef))
+Fetches the entries in the `mri_upload` table that have specific column
```
Comment (cmadjar) [inline docs/scripts_md/MriScanTypeOB.md]: shouldn't this be mri_scan_type table here?

File: docs/scripts_md/MriUploadOB.md [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    } catch(NeuroDB::objectBroker::ObjectBrokerException $e) {
+        die sprintf(
+            "Failed to retrieve MRI uploads: %s",
+            $e->errorMessage
+        );
+    }
+
+# DESCRIPTION
+
+This class provides a set of methods to either fetch or insert mri upload
```
Comment (cmadjar) [inline docs/scripts_md/MriUploadOB.md]: MRI should be all CAP

File: docs/scripts_md/MriUploadOB.md [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+records. The operations are always performed on database table `mri_upload`.
```
Comment (cmadjar) [inline docs/scripts_md/MriUploadOB.md]: should it be "on database table" or "on **the** database table"? I am unsure

File: docs/scripts_md/MriUploadOB.md [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+
+INPUTS:
+    - boolean indicating if only a count of the records found is needed
+      or the full record properties
+    - path of the archive location
+    - boolean indicating if a match is sought on the full archive name
+      or only the basename
+
+RETURNS: a reference to an array of array references. If `$isCount` is true, then
+         `$returnValue-`\[0\]->\[0\]> will contain the count of records sought. Otherwise
```
Comment (cmadjar) [inline docs/scripts_md/MriUploadOB.md]: shouldn't the indexes be in code format as well? I think something is wrong with the way they are formatted

File: docs/scripts_md/ObjectBrokerException.md [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+
+This class is the base class for object broker related exceptions. All object
+brokers should use this class to report unexpected errors/behaviour. You use this
+class by calling the `throw` method with a specific error message (passed
+as argument). This will build a new instance of this class and throw
+it as expected.
+
+### `toString()`
+
+Default string representation of this exception: its associated error
```
Comment (cmadjar) [inline docs/scripts_md/ObjectBrokerException.md]: missing RETURNS section

File: docs/scripts_md/TarchiveOB.md [hunk 22]
```diff
@@ -0,0 +1,95 @@
+# NAME
+
+NeuroDB::objectBroker::TarchiveOB -- An object broker for tarchive records
```
Comment (cmadjar) [inline docs/scripts_md/TarchiveOB.md]: tarchive should be in code format or replace tarchive by "DICOM archive"

File: docs/scripts_md/UnexpectedValueException.md [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+This class is used when an unexpected value is obtained during execution of a command, like
+a database query, reading text from a file, etc... 
+
+### `toString()`
+
+Default representation of this exception when used in a string context.
+Among other things, the returned string will be used for uncaught exceptions
+that make a script die. Note that the returned string can be useful for debugging
+purposes when trying to diagnose why a particular SQL statement did not execute
+successfully.
```
Comment (cmadjar) [inline docs/scripts_md/UnexpectedValueException.md]: missing RETURNS section

File: docs/scripts_md/Database.md (hunk 4 above)
Comment (nicolasbrossard) [reply docs/scripts_md/Database.md]: OK

File: docs/scripts_md/Database.md (hunk 5 above)
Comment (nicolasbrossard) [reply docs/scripts_md/Database.md]: Ditto

File: docs/scripts_md/Database.md (hunk 6 above)
Comment (nicolasbrossard) [reply docs/scripts_md/Database.md]: Done

File: docs/scripts_md/Database.md (hunk 7 above)
Comment (nicolasbrossard) [reply docs/scripts_md/Database.md]: Corrected

File: docs/scripts_md/Database.md (hunk 8 above)
Comment (nicolasbrossard) [reply docs/scripts_md/Database.md]: Absolutely

File: docs/scripts_md/Database.md (hunk 9 above)
Comment (nicolasbrossard) [reply docs/scripts_md/Database.md]: Right

File: docs/scripts_md/Database.md (hunk 10 above)
Comment (nicolasbrossard) [reply docs/scripts_md/Database.md]: Right again

File: docs/scripts_md/Database.md (hunk 11 above)
Comment (nicolasbrossard) [reply docs/scripts_md/Database.md]: I was missing a couple of dots throughout the file actually. I believe I added them all.

File: docs/scripts_md/Database.md (hunk 12 above)
Comment (nicolasbrossard) [reply docs/scripts_md/Database.md]: Yep

File: docs/scripts_md/MriScanTypeOB.md (hunk 14 above)
Comment (nicolasbrossard) [reply docs/scripts_md/MriScanTypeOB.md]: You got it.

Comment (nicolasbrossard) [reply docs/scripts_md/MriScanTypeOB.md]: Right

File: docs/scripts_md/MriScanTypeOB.md (hunk 15 above)
Comment (nicolasbrossard) [reply docs/scripts_md/MriScanTypeOB.md]: Corrected.

File: docs/scripts_md/MriScanTypeOB.md (hunk 16 above)
Comment (nicolasbrossard) [reply docs/scripts_md/MriScanTypeOB.md]: Corrected.

File: docs/scripts_md/MriUploadOB.md (hunk 20 above)
Comment (nicolasbrossard) [reply docs/scripts_md/MriUploadOB.md]: Ok, I'll put the whole thing in code format.

File: docs/scripts_md/ObjectBrokerException.md (hunk 21 above)
Comment (nicolasbrossard) [reply docs/scripts_md/ObjectBrokerException.md]: Ok, added.

File: docs/scripts_md/TarchiveOB.md (hunk 22 above)
Comment (nicolasbrossard) [reply docs/scripts_md/TarchiveOB.md]: Tarchive is now in code format.

File: docs/scripts_md/UnexpectedValueException.md (hunk 23 above)
Comment (nicolasbrossard) [reply docs/scripts_md/UnexpectedValueException.md]: OK, added.

File: docs/scripts_md/Database.md (hunk 13 above)
Comment (nicolasbrossard) [reply docs/scripts_md/Database.md]: Agreed.

File: docs/scripts_md/Database.md [hunk 24]
```diff
@@ -57,42 +57,41 @@ the requests.
...
 name are the empty string, the constructor will call `die`.
 
-INPUTS: a set of properties for the current database:
-
-\- name of the user for the (upcoming) connection
-\- name of the database
-\- name of the host on which the database resides
-\- password for the (upcoming) connection
-\- port used for the (upcoming) connection (defaults to 3306 if not provided)
+INPUTS:
```
Comment (cmadjar) [inline docs/scripts_md/Database.md]: The list of INPUTS in the MD file still looks weird.

File: docs/scripts_md/MRI.md [hunk 25]
```diff
@@ -106,13 +106,13 @@ INPUTS: subjectIDs hashref and database handle reference
...
 
-### identify\_scan\_db($center\_name, $objective, $fileref, $dbhr)
+### identify\_scan\_db($center\_name, $objective, $fileref, $dbhr, $db, $minc\_location)
 
 Determines the type of the scan described by MINC headers based on
 `mri_protocol` table in the database.
 
 INPUTS: center's name, objective of the study, file hashref, database handle
-reference
+reference, database object and location of the minc files.
```
Comment (cmadjar) [inline docs/scripts_md/MRI.md]: MINC should be all CAP

File: docs/scripts_md/MRIProcessingUtility.md [hunk 26]
```diff
@@ -44,13 +44,15 @@ scripts of LORIS.
...
 
-### new($dbhr, $debug, $TmpDir, $logfile, $verbose) >> (constructor)
+### new($db, $dbhr, $debug, $TmpDir, $logfile, $verbose) >> (constructor)
 
 Creates a new instance of this class. The parameter \`\\$dbhr\` is a reference
 to a DBI database handle, used to set the object's database handle, so that all
 the DB-driven methods will work.
 
-INPUT: DBI database handle
+INPUT: database object, DBI database handle, degug flag (1 for debug, 0 otherwise), temporay
```
Comment (cmadjar) [inline docs/scripts_md/MRIProcessingUtility.md]: INPUTS

Comment (cmadjar) [reply docs/scripts_md/MRIProcessingUtility.md]: Probably best to have the INPUTS as a list?

File: docs/scripts_md/MriUploadOB.md (hunk 19 above)
Comment (cmadjar) [reply docs/scripts_md/MriUploadOB.md]: cool!

File: uploadNeuroDB/NeuroDB/Database.pm [hunk 27]
```diff
@@ -72,17 +72,16 @@ use constant DEFAULT_DB_PORT => 3306;
...
 
-INPUTS: a set of properties for the current database:
-
-- name of the user for the (upcoming) connection
-- name of the database
-- name of the host on which the database resides
-- password for the (upcoming) connection
-- port used for the (upcoming) connection (defaults to 3306 if not provided)
+INPUTS:
+- name of the user for the (upcoming) connection.
```
Comment (cmadjar) [inline uploadNeuroDB/NeuroDB/Database.pm]: Maybe add two spaces before the -, it might solve the issue with the MD file later on?

**Comment by nicolasbrossard** [comment]
I rebased this PR and corrected a few things. Should be ready for another review.

**Pull request merged**